src/libs/sfdk/virtualmachine.h

```diff
@@ -0,0 +1,211 @@
+** will be met: http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.
+**
+** Other Usage
+**
+** Alternatively, this file may be used in accordance with the terms and
+** conditions contained in a signed written agreement between you and Digia.
+**
+****************************************************************************/
+
+#pragma once
+
+#include "sfdkglobal.h"
+#include "asynchronous.h"
+
+#include <QUrl>
+
+#include <memory>
+
+namespace QSsh {
+class SshConnectionParameters;
+}
+
+namespace Sfdk {
+
+class SFDK_EXPORT VirtualMachineDescriptor
+{
+public:
+    QUrl uri;
+    QString displayType;
+    QString name;
+};
+
+class VirtualMachinePrivate;
+class SFDK_EXPORT VirtualMachine : public QObject
+{
+    Q_OBJECT
+
+public:
+    class ConnectionUi;
+
+    // Keep in sync with VmConnection::str()
+    enum State {
+        Disconnected,
+        Starting,
+        Connecting,
+        Error,
+        Connected,
+        Disconnecting,
+        Closing
+    };
+    Q_ENUM(State)
+
+    enum ConnectOption {
+        NoConnectOption = 0x00,
+        AskStartVm = 0x01,
+        Block = 0x02,
+    };
+    Q_DECLARE_FLAGS(ConnectOptions, ConnectOption)
+
+    enum Synchronization {
+        Asynchronous,
+        Synchronous
+    };
+
+    ~VirtualMachine() override;
+
+    QUrl uri() const;
+    QString type() const;
+    QString displayType() const;
+    // FIXME add qualifiedName() that would include type information?
+    QString name() const;
+
+    State state() const;
+    QString errorString() const;
+
+    QSsh::SshConnectionParameters sshParameters() const; // FIXME internal
+    void setSshParameters(const QSsh::SshConnectionParameters &sshParameters);
+
+    bool isHeadless() const;
+    void setHeadless(bool headless);
+
+    bool isAutoConnectEnabled() const; // FIXME internal?
+    void setAutoConnectEnabled(bool autoConnectEnabled);
+
+    bool isOff(bool *runningHeadless = 0, bool *startedOutside = 0) const;
+    bool lockDown(bool lockDown);
+    bool isLockedDown() const;
+
+    int memorySizeMb() const;
+    void setMemorySizeMb(int memorySizeMb, const QObject *context,
+            const Functor<bool> &functor);
+    static int availableMemorySizeMb();
+
+    int cpuCount() const;
+    void setCpuCount(int cpuCount, const QObject *context,
+            const Functor<bool> &functor);
+    static int availableCpuCount();
+
+    // FIXME rename as storageCapacityMb
+    int vdiCapacityMb() const;
+    void setVdiCapacityMb(int vdiCapacityMb, const QObject *context,
+            const Functor<bool> &functor);
+
+    bool hasPortForwarding(quint16 hostPort, QString *ruleName = nullptr) const;
+    void addPortForwarding(const QString &ruleName, const QString &protocol,
+            quint16 hostPort, quint16 emulatorVmPort, const QObject *context,
+            const Functor<bool> &functor);
+    void removePortForwarding(const QString &ruleName, const QObject *context,
+            const Functor<bool> &functor);
```

Comment:
It seems these three methods (hasPortForwarding, addPortForwarding and removePortForwarding) are not called anywhere

Reply:
Used by the sensors emulation plugin developed by FRUCT - didn't we already talk about these? :)

Reply:
Maybe we did. Where are the sources for that plugin? Or some interface specification? Or something to help me review this. You don't really expect me to just rubber stamp this code without checking that it does what is expected from it?

Reply:
https://github.com/sailfishos/sailfish-qtcreator/pull/165

Reply:
Thanks. On a closer look I found the original conversation on the .cpp side. And you also linked the fruct in that conversation... So please accept my apologies - I was clearly lazy. And/or been looking at this PR for too long, time to do something else for a change.